bundle/config/mutator/resource_key.go

```diff
@@ -0,0 +1,106 @@
+package mutator
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/databricks/cli/libs/dyn"
+)
+
+// ResourceKey uniquely identifies a resource in configuration.
+type ResourceKey struct {
+	// Type is the type of the resource. E.g. "jobs", "dashboards", etc.
+	Type string
+
+	// Path is the path to the resource.
+	Name string
+}
+
+// ResourceKeySet is a set of resource keys in configuration.
+type ResourceKeySet map[string]map[string]struct{}
+
+func NewResourceKeySet() ResourceKeySet {
+	return make(map[string]map[string]struct{})
+}
+
+// Add adds a resource key to the set.
+func (r ResourceKeySet) AddResourceKey(key ResourceKey) {
+	if _, ok := r[key.Type]; !ok {
+		r[key.Type] = make(map[string]struct{})
+	}
+
+	r[key.Type][key.Name] = struct{}{}
+}
+
+// Add adds resources associated with prefix/value to the set.
+//
+// prefix is a path to value in configuration.
+//
+// If prefix is a resource key ("resources.jobs.job_1"), it's added to the set.
+// If prefix is a resource value ("resources.jobs.job_1.foo"), resource key is extracted and added.
+// If prefix is a sub-tree ("resources" or "resources.jobs"), all resource keys are added.
+func (r ResourceKeySet) Add(prefix dyn.Path, value dyn.Value) error {
+	if len(prefix) >= 3 {
+		parsed, err := GetResourceKey(prefix)
+		if err != nil {
+			return err
+		}
+
+		r.AddResourceKey(parsed)
+
+		return nil
+	} else {
+		if value.Kind() != dyn.KindMap {
+			return fmt.Errorf("expected value to be a map, got %s", value.Kind())
+		}
+
+		for _, pair := range value.MustMap().Pairs() {
+			key := dyn.Key(pair.Key.MustString())
+
+			err := r.Add(prefix.Append(key), pair.Value)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
You can use `dyn.MapByPattern([...] "resources.*.*")` depending on the length of the prefix.

This loop will behave differently for `len(prefix) == 2` vs `len(prefix) == 3`, I believe.

I think it will be easier to grok what happens here by making a switch/case on the length of the prefix, given that depending on the length, the pattern you need to walk on the value is different.

Reply:
Thanks! I've changed API to be pattern-driven and it made it much simpler